### PR TITLE
Creates PyTorch.org doc sources

### PR DESCRIPTION
**Pull request by mruberry**
Updated version of https://github.com/pytorch/xla/pull/1119 per review.

- Updates API_GUIDE to make it more consistent with PyTorch.org CUDA Semantics. It now includes more context on XLA tensors and additional examples.
- Adds PLACEHOLDER doc strings to classes and functions we expect to be public.
- Adds instructions for building html docs from API_GUIDE and index.rst file.
- Adds static HTML content for doc production. 

See https://github.com/pytorch/xla/tree/gh-pages/docs/build for a branch with the build artifacts.

### CODE REVIEW

File: API_GUIDE.md [hunk 1]
```diff
@@ -1,76 +1,118 @@
...
 ```
 
-The XLA device is not a physical device but instead stands in for either a Cloud TPU or CPU. The underlying storage for XLA tensors is a contiguous buffer in device memory and the code in the model shouldn't assume any stride.
+Like other device types, XLA tensors only work with other XLA tensors on the
+same device. So code like
 
-XLA Tensor doesn't support converting single tensor to half precision using `tensor.half()`. Instead, environment variable `XLA_USE_BF16` is available, which converts **all** PyTorch float values to bfloat16 when sending them to the TPU device. The conversion is totally transparent to the user, and the XLA tensors will still retain a float dtype. Similarly, when the tensor is moved back to CPU, its type will be float.
+```python
+l_in = torch.randn(10, device=xm.xla_device())
+linear = torch.nn.Linear(10, 20).cpu()
```
Comment (dlibenzi) [inline API_GUIDE.md]: I think you can omit `.cpu()` here as it is the default.

File: API_GUIDE.md [hunk 2]
```diff
@@ -1,76 +1,118 @@
...
 
-There are different ways to run a model using the PyTorch/XLA framework.
+Building a new PyTorch network or converting an existing one to run on XLA devices requires only a few lines of XLA-specific code. The following snippets highlight these lines when running on a single device, multiple devices with XLA multiprocessing, or multiple threads with XLA multithreading.
 
-### Native PyTorch API
+### Running on a Single XLA Device
 
-The simplest (but not good performing) one is to just run on one core and send the input tensors to the XLA devices manually:
+The following snippet shows a network training on one of several
+device types:
```
Comment (dlibenzi) [inline API_GUIDE.md]: Some lines are really long, some instead seem to wrap around even though they'd fit an 80 column.
Would be nice to decide a standard and stick to it.
IMHO an 80 columns break look nicer on text editors.

File: API_GUIDE.md [hunk 3]
```diff
@@ -1,76 +1,118 @@
...
 
 ```python
 import torch_xla.core.xla_model as xm
 
-device = xm.xla_device()
-model = MNIST()
+device = 'cpu'
```
Comment (dlibenzi) [inline API_GUIDE.md]: I'd honestly leave this on `XLA` and remove the CUDA part.
We are not teaching general PyTorch here, but only differences WRT vanilla PyTorch.
People can connect the dots IMHO.

File: API_GUIDE.md [hunk 4]
```diff
@@ -1,76 +1,118 @@
...
-`optimizer.step()`. This is required because of the way XLA tensors work:
-operations are not executed immediately, but rather added to a graph of pending
-operations which is only executed when its results are required. Using
-`xm.optimizer_step(optimizer, barrier=True)` acts as an execution barrier which forces the
-evaluation of the graph accumulated for a single step. Without this barrier, the
-graph would only be evaluated when evaluating the accuracy of the model, which
-is only done at the end of an epoch, for this example. Even for small models,
-the accumulated graph would be too big to evaluate at the end of an entire
-epoch.
+  if ON_XLA:
```
Comment (dlibenzi) [inline API_GUIDE.md]: Same here.

File: API_GUIDE.md [hunk 5]
```diff
@@ -129,44 +178,79 @@ for epoch in range(1, num_epochs + 1):
...
+Lazy execution is generally invisible to the caller. PyTorxh/XLA automatically
+constructs the graphs, sends them to XLA devices, and synchronizes when
+copying data between an XLA device and the CPU. Inserting a `barrier` when
+taking an optimizer step also synchronizes the CPU and the XLA device.
 
-1. Serialization of XLA tensors doesn't preserve view-relationship.
+### XLA Tensors and bFloat16
 
-   In normal PyTorch devices like CPU/CUDA, view-relationship is preserved when you save & load tensors sharing the same underlying storage.
+When Cloud TPUs are used as XLA devices then:
```
Comment (dlibenzi) [inline API_GUIDE.md]: I'd omit "Cloud' here.

File: API_GUIDE.md [hunk 6]
```diff
@@ -129,44 +178,79 @@ for epoch in range(1, num_epochs + 1):
...
 
-   ``` Python
-   a = torch.rand(3, 3)
-   b = a[0]
-   c = a[0:2]
-   ```
-   That means loaded `b` and `c` still share the same storage. `c` is updated along with `b` and vice versa.
+- Floating point (float and double) XLA tensors use the same datatype
```
Comment (dlibenzi) [inline API_GUIDE.md]: This does not sound clear to me.
We should have one bullet saying that on TPU a both double and float are represented as 32bit floating point.

Then the next bullet should say that if env var is set, then all floating point value will be, once on TPU, represented as bfloat16.
I'd add a link too:

https://en.wikipedia.org/wiki/Bfloat16_floating-point_format

File: API_GUIDE.md [hunk 7]
```diff
@@ -129,44 +178,79 @@ for epoch in range(1, num_epochs + 1):
...
+- If the `XLA_USE_BF16` environment variable is set this datatype is `bfloat16`.
+If not set it's fp32.
 
-   In XLA case, `b` and `c` are separate tensors that one doesn't change with the other.
+XLA tensors on Cloud TPUs will report their PyTorch datatypes regardless of
```
Comment (dlibenzi) [inline API_GUIDE.md]: Remove Cloud. Just TPU. Whether on Cloud or on the ground, does not change the HW 😁

File: docs/source/conf.py [hunk 8]
```diff
@@ -0,0 +1,165 @@
+#!/usr/bin/env python3
```
Comment (dlibenzi) [inline docs/source/conf.py]: It is not clear to me who runs this script ...

Comment (mruberry) [reply docs/source/conf.py]: The script is run by Sphinx. If you run the .sh script it will run Sphinx which runs this.

File: API_GUIDE.md (hunk 2 above)
Comment (mruberry) [reply API_GUIDE.md]: Good catch. Mostly under 80 columns now.

File: API_GUIDE.md (hunk 1 above)
Comment (mruberry) [reply API_GUIDE.md]: Sure.

File: API_GUIDE.md (hunk 2 above)
Comment (mruberry) [reply API_GUIDE.md]: Right.

File: API_GUIDE.md (hunk 3 above)
Comment (mruberry) [reply API_GUIDE.md]: Gone.

File: API_GUIDE.md (hunk 4 above)
Comment (mruberry) [reply API_GUIDE.md]: Yep.

File: API_GUIDE.md (hunk 5 above)
Comment (mruberry) [reply API_GUIDE.md]: Removed all clouds.

File: API_GUIDE.md (hunk 6 above)
Comment (mruberry) [reply API_GUIDE.md]: Updated per feedback.

File: API_GUIDE.md (hunk 7 above)
Comment (mruberry) [reply API_GUIDE.md]: Done.

File: docs/source/conf.py (hunk 8 above)
Comment (mruberry) [reply docs/source/conf.py]: I added a comment at the start of conf.py to this effect.

File: TROUBLESHOOTING.md [hunk 9]
```diff
@@ -119,6 +119,18 @@ If your model shows bad performance, keep in mind the following caveats:
    * When dataset is small, and there are too few steps, this may result in a no-op epoch. Therefore, it is better to use
    small batch sizes in those cases.
 
+## XLA Tensor Quirks
+
+1. **XLA tensor internals are opaque.** XLA tensors always appear to be contiguous and without storage. Networks should not try to check the strides of XLA tensors.
+
+1. **XLA tensors should be moved to the CPU before saving them.** XLA tensors lose their view relationships, if any, when saved. When loaded they are
+placed back on the device they were saved from, but if this device is
+unavailable the load will fail. Moving XLA tensors to the CPU before saving them
```
Comment (ailzhang) [inline TROUBLESHOOTING.md]: Moving tensors across devices actually loses view information. I think we need to rephrase this sentence.

Comment (mruberry) [reply TROUBLESHOOTING.md]: You're right this is very confusing. Updating.
